### PR TITLE
ARROW-11913: [Rust] Improve performance of StringBuilder by delaying bitmap creation

### PR DESCRIPTION
**Pull request by ilya-biryukov**
`PrimitiveBuilder<u8>` was performing a lot of unnecessary bit operations.
One operation for each byte of each string one appends.

Delaying the creation of null bitmap builder for `PrimitiveBuilder`
significantly improves performance of `StringBuilder`. The optimization is
also generally useful as the null bitmap is not stored if resulting
arrays do not contain nulls.

Results from the added benchmark on my machine are:

```
bench_primitive/bench_string
                        time:   [6.8971 ms 6.9123 ms 6.9335 ms]
                        thrpt:  [937.48 MiB/s 940.35 MiB/s 942.43 MiB/s]
                 change:
                        time:   [-49.585% -49.395% -49.204%] (p = 0.00 < 0.05)
                        thrpt:  [+96.867% +97.610% +98.355%]
                        Performance has improved.
```

### CODE REVIEW

File: rust/arrow/src/array/builder.rs [hunk 1]
```diff
@@ -577,14 +579,15 @@ impl<T: ArrowPrimitiveType> PrimitiveBuilder<T> {
...
-        self.bitmap_builder.append(true);
+        self.bitmap_builder.as_mut().map(|b| b.append(true));
         self.values_builder.append(v);
         Ok(())
     }
 
     /// Appends a null slot into the builder
     pub fn append_null(&mut self) -> Result<()> {
-        self.bitmap_builder.append(false);
+        self.materialize_bitmap_builder();
```
Comment (Dandandan) [inline rust/arrow/src/array/builder.rs:589]: Could you also benchmark the case where there are nulls in the input? Wouldn't that be a bit slower now?

**Comment by Dandandan** [comment]
Thanks a lot @ilya-biryukov the builders are not so efficient at the moment indeed.

I think the future directlion of creating data seems to go more in the direction of using iterators / collect() for creating arrow arrays.

Alternatively I think we could have non-null builders which would not create a bitmap and miss the append_null method.

**Comment by ilya-biryukov** [comment]
@Dandandan if the input data is generated from existing Rust datastructures/code, using iterators and collect actually does make a lot of sense. 

I believe our use-case will still be centered around the builders. We're looking at speeding up some `datafusion` operations and an extra round-trip through Rust representation would slow things down. Building with iterators looks hard at times, e.g. efficiently constructing an array of strings is actually much simpler with `StringBuilder` than directly in Rust, as `String` allocates on the heap and this can add up when the Strings are many.

> Alternatively I think we could have non-null builders which would not create a bitmap and leave the append_null method in that case.
Actually I would argue that the current approach makes a lot of sense. Nulls don't add that much runtime overhead and adding new buidlers will probably add up to the complexity. Applying the trick from this commit for builders would bring performance really close to optimal, up to a point that this should almost never become a problem in practice.
Note that `PrimitiveBuilder<u8>` is a somewhat pathological case when used inside `StringBuilder`. Paying for bit-set operation on each byte inside the every added string is just too expensive.

File: rust/arrow/src/array/builder.rs (hunk 1 above)
Comment (ilya-biryukov) [reply rust/arrow/src/array/builder.rs:589]: The change is within the noise threshold. That is exactly what I expected, there is enough code around the buidlers that adding an extra branch for conditionally creating the array does not make much difference.

The benchmark results are:
```
Before:
bench_primitive_nulls/bench_primitive_nulls
                        time:   [4.5728 ms 4.5761 ms 4.5795 ms]

After:
bench_primitive_nulls/bench_primitive_nulls
                        time:   [4.5972 ms 4.6004 ms 4.6038 ms]
                        change: [+0.4255% +0.5290% +0.6284%] (p = 0.00 < 0.05)
                        Change within noise threshold.
```

Comment (Dandandan) [reply rust/arrow/src/array/builder.rs:589]: Thanks, makes sense!

**Comment by Dandandan** [comment]
@ilya-biryukov 
Thanks for updating it with a benchmark. I think this PR is good to go for me as it seems a net win based on the benchmark results. Can you solve the clippy lint?

I am not sure I follow you 100% on using the iterator to build an arrow string array. They are implemented for `&str` as well, so they don't copy the data and should instead translate into a more simple loop. Additionaly, they can pre-allocate the indices array if using an iterator with known size, so they are/should be more efficient than the string builder.
I think we should have a look at the cases where DataFusion is currently using the builder and if possible translate it to use the iterator API.

**Comment by ilya-biryukov** [comment]
@Dandandan 

> @ilya-biryukov
> Thanks for updating it with a benchmark. I think this PR is good to go for me as it seems a net win based on the benchmark results. Can you solve the clippy lint?

Done.

> I am not sure I follow you 100% on using the iterator to build an arrow string array. They are implemented for `&str` as well, so they don't copy the data and should instead translate into a more simple loop.

The use-case I tried to describe involves [building arrays of unknown length](https://github.com/cube-js/arrow/blob/40e2f0e7e596c71ec9016fef9842e2fbff1ff125/rust/datafusion/src/physical_plan/sorted_aggregate.rs#L23) from multiple inputs that get [sent in batches](https://github.com/cube-js/arrow/blob/40e2f0e7e596c71ec9016fef9842e2fbff1ff125/rust/datafusion/src/physical_plan/sorted_aggregate.rs#L69).


I can hardly see anything more convenient than builders here. But this is definitely not the typical use-case, iterator-based APIs are more convenient in most cases.

> Additionaly, they can pre-allocate the indices array if using an iterator with known size, so they are/should be more efficient than the string builder.
> I think we should have a look at the cases where DataFusion is currently using the builder and if possible translate it to use the iterator API.

Note that the code I linked is **not** upstream  DataFusion, it's from our fork. I should have mentioned that explicitly when mentioning DataFusion in the first place to avoid potential confusion.

**Comment by Dandandan** [comment]
Ah, makes sense @ilya-biryukov . I remember seeing cube-js before - looks cool.

How is the forking process working for you, do you keep it updated sometimes?
Are there any changes/features we could upstream to DataFusion (what is the license of those changes?)?

**Review by Dandandan — APPROVED** [review]

**Comment by Dandandan** [comment]
@jorgecarleitao this seems like a nice performance improvement for the builders to lazily allocate the null bitmap. Should also bring some overall performance improvements to DataFusion or readers that use the builder for non-null columns.
What do you think?

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @ilya-biryukov

**Comment by alamb** [comment]
Thanks @ilya-biryukov  -- looks very cool

**Pull request closed without merging**